src/Lecture2.hs

```diff
@@ -1,170 +1,234 @@
+-- most polymorphic type signature you can.
+--
+-- Unlike exercises to Lecture 1, this module also contains more
+-- challenging exercises. You don't need to solve them to finish the
+-- course but you can if you like challenges :)
 module Lecture2
-    ( -- * Normal
-      lazyProduct
-    , duplicate
-    , removeAt
-    , evenLists
-    , dropSpaces
-
-    , Knight (..)
-    , dragonFight
-
-      -- * Hard
-    , isIncreasing
-    , merge
-    , mergeSort
-
-    , Expr (..)
-    , Variables
-    , EvalError (..)
-    , eval
-    , constantFolding
-    ) where
+  ( -- * Normal
+    lazyProduct,
+    duplicate,
+    removeAt,
+    evenLists,
+    dropSpaces,
+    Knight (..),
+    dragonFight,
+
+    -- * Hard
+    isIncreasing,
+    merge,
+    mergeSort,
+    Expr (..),
+    Variables,
+    EvalError (..),
+    eval,
+    constantFolding,
+  )
+where
 
 -- VVV If you need to import libraries, do it after this line ... VVV
+import Data.Char
 
--- ^^^ and before this line. Otherwise the test suite might fail  ^^^
-
-{- | Implement a function that finds a product of all the numbers in
-the list. But implement a lazier version of this function: if you see
-zero, you can stop calculating product and return 0 immediately.
+-- ^ ^^ and before this line. Otherwise the test suite might fail  ^^^
 
->>> lazyProduct [4, 3, 7]
-84
--}
+-- | Implement a function that finds a product of all the numbers in
+-- the list. But implement a lazier version of this function: if you see
+-- zero, you can stop calculating product and return 0 immediately.
+--
+-- >>> lazyProduct [4, 3, 7]
+-- 84
 lazyProduct :: [Int] -> Int
-lazyProduct = error "TODO"
-
-{- | Implement a function that duplicates every element in the list.
-
->>> duplicate [3, 1, 2]
-[3,3,1,1,2,2]
->>> duplicate "cab"
-"ccaabb"
--}
+lazyProduct [] = 1
+lazyProduct (0 : _) = 0
+lazyProduct (h : t) = h * lazyProduct t
+
+-- lazyProduct =
+--   iter id
+--   where
+--     iter k [] = k 1
+--     iter k (h : t) = iter (k . \x -> h * x) t
+
+-- | Implement a function that duplicates every element in the list.
+--
+-- >>> duplicate [3, 1, 2]
+-- [3,3,1,1,2,2]
+-- >>> duplicate "cab"
+-- "ccaabb"
 duplicate :: [a] -> [a]
-duplicate = error "TODO"
+duplicate =
+  let dupleIter [] = []
+      dupleIter (h : t) = h : h : dupleIter t
+   in dupleIter
+
+-- | Implement function that takes index and a list and removes the
+-- element at the given position. Additionally, this function should also
+-- return the removed element.
+--
+-- >>> removeAt 0 [1 .. 5]
+-- (Just 1,[2,3,4,5])
+--
+-- >>> removeAt 10 [1 .. 5]
+-- (Nothing,[1,2,3,4,5])
+removeAt :: Integer -> [a] -> (Maybe a, [a])
+removeAt =
+  removeIter (Nothing, [])
+  where
+    removeIter (x, l) _ [] = (x, reverse l)
+    removeIter (x, l) count (h : t)
+      | count < 0 = (Nothing, h : t)
+      | count == 0 = (Just h, reverse l ++ t)
+      | otherwise = removeIter (x, h : l) (count - 1) t
```

Comment:
This is a valid implementation 👍🏻 

It has a few inefficient places:

* You compare `count` with `< 0` on every recursive step where it's enough to compare only once
* Reversing the list at the end

You can implement this function without these two drawbacks in the following more efficient way:

```suggestion
removeAt :: Integer -> [a] -> (Maybe a, [a])
removeAt n l
    | n < 0 = (Nothing, l)
    | otherwise = go n l
  where
    go _ [] = (Nothing, [])
    go 0 (x : xs) = (Just x, xs)
    go count (x : xs) =
        let (val, rest) = go (count - 1) xs
        in (val, x : rest)
```